### ISSUE TITLE
Naming/Javadoc suggestions for Retry.java

### ISSUE THREAD

**Issue by mickeelm** [reactor/reactor-core#2553]
## Documentation Issue

I think that both the naming and the javadoc regarding number of retries in `Retry.java` could use some changes for clarity.

* Naming. `totalRetries` is great, but `totalRetriesInARow` is, IMHO, if not confusing then not greatly named. My suggestion is to deprecate it and change to `consecutiveRetries`.
* I think the javadoc part `ZERO BASED...` should be omitted. The gist of `can also be read as...` is perfectly sufficient, keep it simple and the word retries ("re" being the key) cannot mean anything else. Also, `a 0-index for the error...` is potentially misleading given *the* error - I first read it as a count of errors with a specific kind of `Throwable`, as if there were parallell counts (one per Throwable-type or something like that) going on.

## Improvement Suggestion

My suggestions for the javadoc of those methods:

```
Returns: the total number of retries since the source first was subscribed to.
```

and

```
Returns: the number of retries since the latest onNext(), or since the source first was subscribed to if there hasn't been any onNext() signals (in which case totalRetries() and consecutiveRetries() are equivalent).
```

Also I suggest that both the parts `that might result in a retry attempt` and `that potentially triggers a retry` (in `generateCompanion()` and `RetrySignal`) be removed. I understand them, but I think it _can_ be misleading, it might be interpreted as if the signals themselves would trigger retries given state x or y. It should be clear that the signal-flux is just a provider of signals/metadata and that it's up to the companion to react to it and decide if there should be any retries or not (even if one decides to return the flux itself, but that's probably not the most common use-case).

Happy to submit a PR if you like any of it.

**Comment by simonbasle** [comment]
hi @mickeelm. At the time I tried to come up with a meaningful name, always a hard endeavor... the name has to stay for at least a while though, now that it has been (recently) released and part of the public API. we can definitely rework the javadoc and a PR with your suggestions would be very welcome there.

Keep in mind `RetrySignal` was added in `3.3.x` so use that branch as the basis for your PR.

**Comment by mickeelm** [comment]
I hear you, naming is super hard. I'll prepare a PR with javadoc suggestions!

**Comment by mickeelm** [comment]
Just a question here...in 3.4.2 I can see `retryContextView()` and "reset/rollbacked". I would express this "reset/rolled back". But, I should only consider the 3.3.x...line if I understand it correctly?

**Comment by simonbasle** [comment]
the way we work is we make a change in the oldest maintenance branch in which we want the fix (here `3.3.x`) and then at merge time a maintainer will forward the fix to newer branches by way of merging. Here we will merge `3.3.x` into `master`, "forwarding" the fix to the 3.4.x line.

if there are subtelties in the merge (due to architectural changes in the newest branch for instance) or if there are further changes to affect in the newest branch, let us know in the PR. 2 choices there: if small we can make the change as we forward-merge. Otherwise the best choice is probably a small follow-up PR on `master` afterwards, referencing either this issue or the original PR in the commit title (eg. "Polish #2553 on 3.4: reset/rollbacked"...)

**Comment by mickeelm** [comment]
Great, I'll just mention it as a comment then, as it's so small. Thanks!

**Comment by mickeelm** [comment]
Allright, one thing that's not clear to me - which one is considered the companion publisher?

1) The incoming `Flux` of `RetrySignal`(s)
2) The returned `Publisher<?>` (which can very well be the incoming "under the hood", if that's just returned as is)
3) Both..?

I mean, the `generateCompanion()`-method would IMO suggest 2), but in the `Retry` javadoc it says "given a companion `Flux` of `Retry.RetrySignal`" which would suggest 1). Which is it? :)


Another thing, this part in `generateCompanion()`:

"Any `onNext` emitted by this publisher will trigger a retry"

is that really true? If I were to create a publisher that just triggers an `onNext` every second even if there hasn't been any errors, wouldn't those signals be supressed (like the behavior of `repeatWhen`)?

**Comment by simonbasle** [comment]
It is indeed a little bit of both. The one that really matters from the user's perspective is the one returned by `generateCompanion()` since that's the one they write to drive the retries. The input to that function is in itself a valid companion, which is intended as the _base_ for the companion ultimately used.

In 99% of cases, we expect that the generated companion is somehow derived from the base companion (and as such, it doesn't generate _more_ signals than the base). Typically, it could shift the signals in time, or terminate earlier than the base with an error (because if eg. the source continuously errors, then the base companion would continuously produce `onNext` representing the errors).

The end-result companion is requested one-by-one in reply to errors in the source, so even if it not derived from the base, it shouldn't matter. But a timer-type companion is definitely a bad choice, because it will cause backpressure errors.

So yeah `"Any onNext emitted by this publisher will trigger a retry"` is a bit misleading, it's only valid "in response to a request of one".
Or maybe rephrase to something along the lines of `each time an error happens in the source, the companion publisher is requested for one signal. An onNext will trigger a retry, while onError fails the outer `Flux` (and `onComplete` swallows the error by completing the `Flux`)."

**Comment by simonbasle** [comment]
This can be observed with the following test:

```java
	@Test
	void unrelatedCompanionBehavior() {
		AtomicInteger subCount = new AtomicInteger();
		Flux.defer(() -> {
			int sc = subCount.incrementAndGet();
			if (sc < 30) {
				return Flux.error(new IllegalStateException("error #" + sc));
			}
			return Flux.just(sc);
		})
		    .retryWhen(Retry.from(companion -> Flux.range(1, 10)
		                                           .concatWith(Mono.error(new IllegalStateException("STOP RETRY")))
		                                           .log()))
		    .as(StepVerifier::create)
		    .expectErrorMessage("STOP RETRY")
		    .verify(Duration.ofSeconds(1));

		assertThat(subCount).hasValue(3);
	}
```

**Comment by mickeelm** [comment]
Thank you for a great answer - I'll try to make it as clear as possible.

**Comment by mickeelm** [comment]
Did I miss the milestone with 2 minutes? :D

**Comment by simonbasle** [comment]
Eh, missed your submission 😱 Today is the last day I can review PRs and the day is already well-advance in Europe. That said, review shouldn't be too hard so there's a good chance for it to be shipped still. I'll have time to review in about 40min, keep an eye out for comments 😉

**Comment by mickeelm** [comment]
I believe we are in the same time zone (I'm from Sweden). I believe I literally submitted the PR two minutes after your milestone change, so it was just funny :)

Rest assured that I don't mind if this is shipped in a later release, just happy to contribute. So, no rush at all!

**Comment by simonbasle** [comment]
All good, easy enough 😄

**Comment by mickeelm** [comment]
Awesome! Happy you liked it :)
